ext/opcache/jit/tls/zend_jit_tls_aarch64.c

```diff
@@ -0,0 +1,247 @@
+#define AARCH64_MOVK_HW_MASK           0x00600000 /* bits 22-21 */
+#define AARCH64_NOP                    0xd503201f
+
+#undef USE_FALLBACK
+
+#ifdef __MUSL__
+
+# define DTV_OFFSET -8
+# define DTV_INDEX_GAP 0
+
+typedef struct _dtv_pointer_t {
+	uintptr_t val;
+} dtv_pointer_t;
+
+typedef struct _tls_descriptor {
+	size_t index;
+	size_t offset;
+} tls_descriptor;
+
+#elif defined(__FreeBSD__)
+
+# define DTV_OFFSET 0
+/* Index is offset by 1 on FreeBSD (https://github.com/freebsd/freebsd-src/blob/22ca6db50f4e6bd75a141f57cf953d8de6531a06/lib/libc/gen/tls.c#L88) */
+# define DTV_INDEX_GAP 1
+
+typedef struct _dtv_pointer_t {
+	uintptr_t val;
+} dtv_pointer_t;
+
+/* https://github.com/freebsd/freebsd-src/blob/c52ca7dd09066648b1cc40f758289404d68ab886/libexec/rtld-elf/aarch64/reloc.c#L180-L184 */
+typedef struct _tls_descriptor {
+	void*   thunk;
+	int     index;
+	size_t  offset;
+} tls_descriptor;
+
+#elif defined(__GLIBC__)
+
+# define DTV_OFFSET 0
+# define DTV_INDEX_GAP 0
+
+typedef struct _dtv_pointer_t {
+	uintptr_t val;
+	uintptr_t _;
+} dtv_pointer_t;
+
+typedef struct _tls_descriptor {
+	size_t index;
+	size_t offset;
+} tls_descriptor;
+
+#else
+# define USE_FALLBACK 1
+#endif
+
+zend_result zend_jit_resolve_tsrm_ls_cache_offsets(
+	size_t *tcb_offset,
+	size_t *module_index,
+	size_t *module_offset
+) {
+#ifdef USE_FALLBACK
+	return FAILURE;
+#else
+	*tcb_offset = tsrm_get_ls_cache_tcb_offset();
+	if (*tcb_offset != 0) {
+		return SUCCESS;
+	}
+
+	void *addr;
+	uint32_t *insn;
+	void *thread_pointer;
+
+	__asm__ __volatile__(
+		/* Load thread pointer address */
+		"mrs   %0, tpidr_el0\n"
+		/* Load next instruction address */
+		"adr   %1, .+4\n\t"
+		/* General Dynamic code sequence as expected by linkers */
+		"adrp  x0, :tlsdesc:_tsrm_ls_cache\n"
+		"ldr   x1, [x0, #:tlsdesc_lo12:_tsrm_ls_cache]\n"
+		"add   x0, x0, :tlsdesc_lo12:_tsrm_ls_cache\n"
+		".tlsdesccall    _tsrm_ls_cache\n"
+		"blr   x1\n"
+		"mrs   x8, tpidr_el0\n"
+		"add   %2, x8, x0\n"
+		: "=r" (thread_pointer), "=r" (insn), "=r" (addr)
+		:
+		: "x0", "x1", "x8");
+
+	ZEND_ASSERT(addr == &_tsrm_ls_cache);
+
+	/* Check if the general dynamic code was relaxed by the linker */
+
+	// adrp x0, #any
+	if ((insn[0] & ~AARCH64_ADRP_IMM_MASK) != 0x90000000) {
+		zend_accel_error(ACCEL_LOG_DEBUG, "adrp insn does not match: 0x%08" PRIx32 "\n", insn[0]);
+		goto code_changed;
+	}
+
+	// ldr x1, [x0, #any]
+	if ((insn[1] & ~AARCH64_LDR_UNSIGNED_IMM_MASK) != 0xf9400001) {
+		zend_accel_error(ACCEL_LOG_DEBUG, "ldr insn does not match: 0x%08" PRIx32 "\n", insn[1]);
+		goto code_changed;
+	}
+
+	// add x0, x0, any
+	if ((insn[2] & ~AARCH64_ADD_IMM_MASK) != 0x91000000) {
+		zend_accel_error(ACCEL_LOG_DEBUG, "add insn does not match: 0x%08" PRIx32 "x\n", insn[2]);
+		goto code_changed;
+	}
+
+	/* Code is intact, we can extract immediate values */
+
+	uint64_t adrp_imm = (uint64_t)( ((insn[0] & 0x00ffffe0) >> 3) | ((insn[0] & 0x60000000) >> 29) ) << 12;
```

Comment:
Why shift the adrp instruction immediate with 3 instead of 5? The immediate bits start at 5?